neqo-transport/src/frame.rs

```diff
@@ -18,68 +18,152 @@ use crate::{
-        match self {
-            Self::Transport(_) => 0,
-            Self::Application(_) => 1,
+impl From<FrameType> for u8 {
+    fn from(val: FrameType) -> Self {
+        val as Self
+    }
+}
+
+impl TryFrom<u64> for FrameType {
+    type Error = Error;
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        match value {
+            0x0 => Ok(Self::Padding),
+            0x1 => Ok(Self::Ping),
+            0x2 => Ok(Self::Ack),
+            0x3 => Ok(Self::AckEcn),
+            0x4 => Ok(Self::ResetStream),
+            0x5 => Ok(Self::StopSending),
+            0x6 => Ok(Self::Crypto),
+            0x7 => Ok(Self::NewToken),
+            0x8 => Ok(Self::Stream),
+            0x9 => Ok(Self::StreamWithFin),
+            0xa => Ok(Self::StreamWithLen),
+            0xb => Ok(Self::StreamWithLenFin),
+            0xc => Ok(Self::StreamWithOff),
+            0xd => Ok(Self::StreamWithOffFin),
+            0xe => Ok(Self::StreamWithOffLen),
+            0xf => Ok(Self::StreamWithOffLenFin),
+            0x10 => Ok(Self::MaxData),
+            0x11 => Ok(Self::MaxStreamData),
+            0x12 => Ok(Self::MaxStreamsBiDi),
+            0x13 => Ok(Self::MaxStreamsUniDi),
+            0x14 => Ok(Self::DataBlocked),
+            0x15 => Ok(Self::StreamDataBlocked),
+            0x16 => Ok(Self::StreamsBlockedBiDi),
+            0x17 => Ok(Self::StreamsBlockedUniDi),
+            0x18 => Ok(Self::NewConnectionId),
+            0x19 => Ok(Self::RetireConnectionId),
+            0x1a => Ok(Self::PathChallenge),
+            0x1b => Ok(Self::PathResponse),
+            0x1c => Ok(Self::ConnectionCloseTransport),
+            0x1d => Ok(Self::ConnectionCloseApplication),
+            0x1e => Ok(Self::HandshakeDone),
+            0xaf => Ok(Self::AckFrequency),
+            0x30 => Ok(Self::Datagram),
+            0x31 => Ok(Self::DatagramWithLen),
+            _ => Err(Error::UnknownFrameType),
         }
     }
+}
 
-    const fn from_type_bit(bit: u64, code: u64) -> Self {
-        if (bit & 0x01) == 0 {
-            Self::Transport(code)
-        } else {
-            Self::Application(code)
+impl FrameType {
+    const fn is_stream_with_length(self) -> bool {
+        matches!(
+            self,
+            Self::StreamWithLen
+                | Self::StreamWithLenFin
+                | Self::StreamWithOffLen
+                | Self::StreamWithOffLenFin
+        )
+    }
+
+    const fn is_stream_with_offset(self) -> bool {
+        matches!(
+            self,
+            Self::StreamWithOff
+                | Self::StreamWithOffFin
+                | Self::StreamWithOffLen
+                | Self::StreamWithOffLenFin
+        )
+    }
+    const fn is_stream_with_fin(self) -> bool {
+        matches!(
+            self,
+            Self::StreamWithFin
+                | Self::StreamWithLenFin
+                | Self::StreamWithOffFin
+                | Self::StreamWithOffLenFin
+        )
```

Comment:
I wonder whether this produces the same code.  (i.e., does the compiler know how to apply `self as u64 & mask == pattern`)

It would be a shame if the cleaner code here results in worse performance.  Do you think that it is worth checking?

Reply:
It doesn't seem to be much worse at least: https://rust.godbolt.org/z/z3o1aMcnv

Reply:
Yep, I tried a few variations and they all seem good.  But the code is neater this way.  Wins all around.